### PR TITLE
Many functions added - and one error corrected?

### PR DESCRIPTION
**Pull request by clavisound**
First of all: thank you. I had success with your library instead of Adafruit and SparkFun.

I made some changes. Maybe you want to merge and continue together?

I added:
1. Debug on write.

Functions (TESTED)
2. readTap
3. readEvent
4. autoSleep
5. configTap

Enhancement
1. added polarity to intConf (I think this is wrong decision, maybe it's better to go in different function)
2. option to disable INT with INT_0 - intConf
3. hardcoded latch (TODO with option)

Functions for experiemnts (UNTESTED)
1. wakeUpInertial (UNTESTED),
2. setClickAda (UNTESTED copy from adafruit - old driver)
3. motionSTforum (UNTESTED),

I think this code is ok
```
if(moveType == 0)       dataToWrite |= 0x0A; // b1010 phone on table
if(moveType == 1)       dataToWrite |= 0x05; // b0101 phone upside down on table
if(moveType == 2)       dataToWrite = 0b00000011; // bike fall
```

You had
```
if(moveType == 1)       dataToWrite |= 0x0A; // b1010 phone on table
if(moveType == 2)       dataToWrite |= 0x05; // b0101 phone upside down on table
```

### CODE REVIEW

File: src/lis3dh-motion-detection.cpp [hunk 1]
```diff
@@ -191,7 +191,64 @@ status_t LIS3DH::writeRegister(uint8_t offset, uint8_t dataToWrite) {
     returnError = IMU_HW_ERROR;
   }
 
+	_DEBBUG("Write register 0x", offset, " = ", dataToWrite);
 	return returnError;
+
+	delay(100);             // some users reported hangs without delay.
```
Comment (ldab) [inline src/lis3dh-motion-detection.cpp:197]: Can you please refer back to the users who reported hangs? I would avoid delay at any cost.

File: src/lis3dh-motion-detection.cpp [hunk 2]
```diff
@@ -191,7 +191,64 @@ status_t LIS3DH::writeRegister(uint8_t offset, uint8_t dataToWrite) {
...
+}
+
+// Read tap (aka: 'click')
+uint8_t LIS3DH::readTap(){
```
Comment (ldab) [inline src/lis3dh-motion-detection.cpp:201]: Tap? Do you refer to Click? I would prefer to use the datasheet nomenclature, i.e Click.

File: src/lis3dh-motion-detection.cpp [hunk 3]
```diff
@@ -191,7 +191,64 @@ status_t LIS3DH::writeRegister(uint8_t offset, uint8_t dataToWrite) {
...
+	readRegister(&t, LIS3DH_INT1_SRC);
+	return t;
+}
+
+void LIS3DH::configTap(bool Zdouble, bool Zsingle, bool Ydouble, bool Ysingle, bool Xdouble, bool Xsingle, uint8_t threshold){
+	uint8_t r = 0;
+
+	writeRegister(LIS3DH_CTRL_REG3, 0x80); 				// Tap on INT1
+	readRegister(&r, LIS3DH_CTRL_REG5);
+	writeRegister(LIS3DH_CTRL_REG5, r | 0b1000); 			// latch INT1
```
Comment (ldab) [inline src/lis3dh-motion-detection.cpp:220]: Why not leave the user to determine if latch is required or not?

I believe we're over complicating the library, if a user needs latching and read the register and source of the interrupts, the user is probably experienced and therefore would not use this library.

File: src/lis3dh-motion-detection.cpp [hunk 4]
```diff
@@ -191,7 +191,64 @@ status_t LIS3DH::writeRegister(uint8_t offset, uint8_t dataToWrite) {
...
+
+	readRegister(&r, LIS3DH_CTRL_REG2);
+	writeRegister(LIS3DH_CTRL_REG2, r | 0x04); 			// Hi pass enable
```
Comment (ldab) [inline src/lis3dh-motion-detection.cpp:223]: Can you please ellaborate what advantage did you see when using the filter?

File: src/lis3dh-motion-detection.cpp [hunk 5]
```diff
@@ -191,7 +191,64 @@ status_t LIS3DH::writeRegister(uint8_t offset, uint8_t dataToWrite) {
...
+	writeRegister(LIS3DH_TIME_WINDOW, 0x40);			// arbitrary
+}
+
+void LIS3DH::autoSleep(uint8_t threshold, uint8_t time){
+	writeRegister(LIS3DH_ACT_THS, threshold);
+	writeRegister(LIS3DH_ACT_DUR, time);
+}
+
+// p. 23 of DOCid1898
+void LIS3DH::wakeUpInertialAN(){
```
Comment (ldab) [inline src/lis3dh-motion-detection.cpp:238]: If untested, please remove it or put it a .dev branch.

File: src/lis3dh-motion-detection.cpp [hunk 6]
```diff
@@ -345,33 +403,48 @@ void LIS3DH::applySettings( void )
...
 	//Build INT_CFG 0x30 or 0x34
 	//Detect movement or stop
-	if(moveType == 1)	dataToWrite |= 0x0A;
-	else 							dataToWrite |= 0x05;
+	//bike straight  : ZL YH XL
+	//bike fall right: ZL YH XH
+	//bile fall left : ZL YL XH
+	if(moveType == 0)	dataToWrite |= 0x0A; // b1010 phone on table
+	if(moveType == 1)	dataToWrite |= 0x05; // b0101 phone upside down on table
+	if(moveType == 2)	dataToWrite = 0b00000011; // bike fall
```
Comment (ldab) [inline src/lis3dh-motion-detection.cpp:428]: This does not make sense, can you please elaborate?

XHIE Enable interrupt generation on X high event or on Direction recognition.
XLIE Enable interrupt generation on X low event or on Direction recognition.

Please remove the comments which are only relevant to your bike application.

File: src/lis3dh-motion-detection.h [hunk 7]
```diff
@@ -68,12 +70,13 @@ typedef enum
...
 	DET_STOP,
 	DET_MOVE,
-	//...
+	BIKE_FALL,
 } event_t;
 
 typedef enum
 {
-	INT_1 = 1,
+	INT_0,
```
Comment (ldab) [inline src/lis3dh-motion-detection.h:78]: INT_0 is not intuitive, I believe it's better to create another function to disable interupts

**Review by ldab — CHANGES_REQUESTED** [review]

File: src/lis3dh-motion-detection.cpp (hunk 1 above)
Comment (clavisound) [reply src/lis3dh-motion-detection.cpp:197]: See last message https://community.st.com/s/question/0D50X00009XkgOnSAJ/lis3dh-inertial-interruptwakeup-not-detected

Also in documentation, in self test they rote that they wait approx 90ms for reading to calm down.

File: src/lis3dh-motion-detection.cpp (hunk 2 above)
Comment (clavisound) [reply src/lis3dh-motion-detection.cpp:201]: Yes, as stated. No problem, I can change it.

File: src/lis3dh-motion-detection.cpp (hunk 3 above)
Comment (clavisound) [reply src/lis3dh-motion-detection.cpp:220]: You have right, I thought it was a sane default. Well the library has many holes.p. I will add it

(I don’t understood the last argument, but no worries)

File: src/lis3dh-motion-detection.cpp (hunk 4 above)
Comment (clavisound) [reply src/lis3dh-motion-detection.cpp:223]: If I remember correctly, it was suggested in an example from the manufacturer.

File: src/lis3dh-motion-detection.cpp (hunk 5 above)
Comment (clavisound) [reply src/lis3dh-motion-detection.cpp:238]: Ok

File: src/lis3dh-motion-detection.cpp (hunk 6 above)
Comment (clavisound) [reply src/lis3dh-motion-detection.cpp:428]: Yes, mistake :)

I don’t think comments where annoying. As you wish.

File: src/lis3dh-motion-detection.h (hunk 7 above)
Comment (clavisound) [reply src/lis3dh-motion-detection.h:78]: Ok

**Comment by clavisound** [comment]
Thank you for the detailed comments.

I retried today the library and it was a disaster. Nothing works as expected, even with your untouched library. More investigation todo. Sorry for the hassle.

Hope will come back. Stay safe :)

**Comment by ldab** [comment]
Thanks for the comments, i have pushed some changes,, please make sure you work on the latest :)

**Comment by clavisound** [comment]
Problem solved, the fault probably was with Arduino IDE. To my surpise 1.8.3 (old) version was producing strange behaviour of LIS3DH (only!). Upgraded to 1.8.12 and it seems fine.

I saw that polarity "passed", thank you.

**Pull request closed without merging**